File: frontend/src/types/observability.ts
export interface SystemHealthItem {
  api_status: string
  readiness_status: string
  backend_status: string
  uptime_seconds: number
  process_id: number
  active_threads: number
  cpu_percent: number
  system_cpu_percent: number
  memory_rss_mb: number
  system_memory_percent: number
  environment: string
  timestamp: string
}

export interface RuntimeMetricsResponse {
  total_requests: number
  successful_requests: number
  error_requests: number
  error_rate_pct: number
  rps: number
  min_latency_ms: number
  p50_latency_ms: number
  p90_latency_ms: number
  p95_latency_ms: number
  p99_latency_ms: number
  max_latency_ms: number
  sample_count: number
  active_window_seconds: number
  has_runtime_data: boolean
}

export interface ForecastOperationsMetrics {
  total_forecasts: number
  successful_forecasts: number
  rejected_forecasts: number
  failed_forecasts: number
  success_rate_pct: number
  rejection_rate_pct: number
  forecasts_by_crop: Record<string, number>
  forecasts_by_strategy: Record<string, number>
  forecasts_by_status: Record<string, number>
  recent_forecast_count: number
  has_runtime_data: boolean
}

export interface StrategyUsageItem {
  crop: string
  strategy: string
  certification_status: string
  algorithm: string
  runtime_invocations_count: number
  runtime_percentage: number | null
  fallback_invocations_count: number
  last_used_timestamp: string | null
}

export interface StrategyMonitoringResponse {
  total_strategies_monitored: number
  total_runtime_observations: number
  has_runtime_data: boolean
  strategies: StrategyUsageItem[]
}

export interface TraceStageItem {
  stage_name: string
  status: string
  duration_ms: number | null
  timestamp?: string | null
  details?: Record<string, any> | null
}

export interface PredictionTraceResponse {
  request_id: string
  timestamp: string
  crop: string
  state: string
  district: string
  forecast_year: number
  strategy?: string | null
  model_name?: string | null
  model_version?: string | null
  prediction?: number | null
  unit: string
  status: string
  error_code?: string | null
  error_message?: string | null
  total_duration_ms?: number | null
  stages: TraceStageItem[]
  provenance_hash?: string | null
  audit_status: string
  model_hash_verified: boolean
  dataset_verified: boolean
}

export interface ModelIntegrityItem {
  model_id: string
  crop: string
  algorithm: string
  version: string
  artifact_path: string
  file_exists: boolean
  registered_sha256: string
  actual_sha256: string
  integrity_status: string
  last_checked_timestamp: string
}

export interface ModelIntegrityResponse {
  total_models_registered: number
  verified_models_count: number
  failed_models_count: number
  overall_integrity_status: string
  models: ModelIntegrityItem[]
}

export interface DatasetIntegrityResponse {
  dataset_name: string
  dataset_version: string
  file_path: string
  file_exists: boolean
  total_records: number
  total_columns: number
  file_size_bytes: number
  sha256_checksum: string
  date_coverage: string
  state_count: number
  district_count: number
  crop_count: number
  schema_status: string
  last_verified_timestamp: string
}

export interface StrategyRegistryHealthResponse {
  registry_path: string
  registry_available: boolean
  compiled_at: string
  total_strategies_registered: number
  production_ready_count: number
  conditional_production_count: number
  baseline_production_count: number
  coverage_records_count: number
  certification_guard_status: string
  overall_status: string
}

export interface OperationalEventItem {
  timestamp: string
  severity: string
  event_type: string
  request_id?: string | null
  endpoint?: string | null
  status_code?: number | null
  message: string
  details?: Record<string, any> | null
}

export interface OperationalErrorsResponse {
  total_events_logged: number
  total_errors_count: number
  recent_events: OperationalEventItem[]
  errors_by_category: Record<string, number>
}

export interface AlertConditionItem {
  alert_id: string
  alert_name: string
  severity: string
  metric_name: string
  current_value: any
  threshold_value: any
  condition: string
  is_active: boolean
  message: string
  timestamp: string
}

export interface AlertsResponse {
  active_alerts_count: number
  active_alerts: AlertConditionItem[]
  resolved_alerts: AlertConditionItem[]
  configured_thresholds: Record<string, any>
}

export interface DriftMonitoringResponse {
  monitoring_notice: string
  overall_drift_status: string
  total_features_evaluated: number
  stable_features_count: number
  drifted_features_count: number
  features: Array<Record<string, any>>
}

export interface ObservabilitySummaryResponse {
  system_health: SystemHealthItem
  runtime_metrics: RuntimeMetricsResponse
  forecast_operations: ForecastOperationsMetrics
  model_integrity_status: string
  dataset_integrity_status: string
  strategy_registry_status: string
  active_alerts_count: number
  timestamp: string
}

File: backend/schemas/observability.py
"""
Pydantic Schemas for Day 28 Production Observability & Operational Intelligence.
"""

from typing import List, Dict, Any, Optional
from pydantic import BaseModel, Field


class SystemHealthItem(BaseModel):
    api_status: str = Field(..., description="API HTTP status (HEALTHY / DEGRADED / UNHEALTHY)")
    readiness_status: str = Field(..., description="Readiness probe status (READY / NOT_READY)")
    backend_status: str = Field(..., description="Backend engine status (OPERATIONAL / DEGRADED)")
    uptime_seconds: float = Field(..., description="Backend process uptime in seconds")
    process_id: int = Field(..., description="Operating system PID")
    active_threads: int = Field(..., description="Active thread count")
    cpu_percent: float = Field(..., description="Current process CPU utilization %")
    system_cpu_percent: float = Field(..., description="Current system-wide CPU utilization %")
    memory_rss_mb: float = Field(..., description="Process resident memory usage in MB")
    system_memory_percent: float = Field(..., description="Host system memory utilization %")
    environment: str = Field(..., description="Deployment environment (LOCAL / CONTAINERIZED / DEVELOPMENT)")
    timestamp: str = Field(..., description="ISO-8601 status timestamp")


class RuntimeMetricsResponse(BaseModel):
    total_requests: int
    successful_requests: int
    error_requests: int
    error_rate_pct: float
    rps: float
    min_latency_ms: float
    p50_latency_ms: float
    p90_latency_ms: float
    p95_latency_ms: float
    p99_latency_ms: float
    max_latency_ms: float
    sample_count: int
    active_window_seconds: float
    has_runtime_data: bool


class ForecastOperationsMetrics(BaseModel):
    total_forecasts: int
    successful_forecasts: int
    rejected_forecasts: int
    failed_forecasts: int
    success_rate_pct: float
    rejection_rate_pct: float
    forecasts_by_crop: Dict[str, int]
    forecasts_by_strategy: Dict[str, int]
    forecasts_by_status: Dict[str, int]
    recent_forecast_count: int
    has_runtime_data: bool


class StrategyUsageItem(BaseModel):
    crop: str
    strategy: str
    certification_status: str
    algorithm: str
    runtime_invocations_count: int
    runtime_percentage: Optional[float] = None
    fallback_invocations_count: int
    last_used_timestamp: Optional[str] = None


class StrategyMonitoringResponse(BaseModel):
    total_strategies_monitored: int
    total_runtime_observations: int
    has_runtime_data: bool
    strategies: List[StrategyUsageItem]


class TraceStageItem(BaseModel):
    stage_name: str
    status: str
    duration_ms: Optional[float] = None
    timestamp: Optional[str] = None
    details: Optional[Dict[str, Any]] = None


class PredictionTraceResponse(BaseModel):
    request_id: str
    timestamp: str
    crop: str
    state: str
    district: str
    forecast_year: int
    strategy: Optional[str] = None
    model_name: Optional[str] = None
    model_version: Optional[str] = None
    prediction: Optional[float] = None
    unit: str = "kg/ha"
    status: str
    error_code: Optional[str] = None
    error_message: Optional[str] = None
    total_duration_ms: Optional[float] = None
    stages: List[TraceStageItem]
    provenance_hash: Optional[str] = None
    audit_status: str
    model_hash_verified: bool
    dataset_verified: bool


class ModelIntegrityItem(BaseModel):
    model_id: str
    crop: str
    algorithm: str
    version: str
    artifact_path: str
    file_exists: bool
    registered_sha256: str
    actual_sha256: str
    integrity_status: str
    last_checked_timestamp: str


class ModelIntegrityResponse(BaseModel):
    total_models_registered: int
    verified_models_count: int
    failed_models_count: int
    overall_integrity_status: str
    models: List[ModelIntegrityItem]


class DatasetIntegrityResponse(BaseModel):
    dataset_name: str
    dataset_version: str
    file_path: str
    file_exists: bool
    total_records: int
    total_columns: int
    file_size_bytes: int
    sha256_checksum: str
    date_coverage: str
    state_count: int
    district_count: int
    crop_count: int
    schema_status: str
    last_verified_timestamp: str


class StrategyRegistryHealthResponse(BaseModel):
    registry_path: str
    registry_available: bool
    compiled_at: str
    total_strategies_registered: int
    production_ready_count: int
    conditional_production_count: int
    baseline_production_count: int
    coverage_records_count: int
    certification_guard_status: str
    overall_status: str


class OperationalEventItem(BaseModel):
    timestamp: str
    severity: str
    event_type: str
    request_id: Optional[str] = None
    endpoint: Optional[str] = None
    status_code: Optional[int] = None
    message: str
    details: Optional[Dict[str, Any]] = None


class OperationalErrorsResponse(BaseModel):
    total_events_logged: int
    total_errors_count: int
    recent_events: List[OperationalEventItem]
    errors_by_category: Dict[str, int]


class AlertConditionItem(BaseModel):
    alert_id: str
    alert_name: str
    severity: str
    metric_name: str
    current_value: Any
    threshold_value: Any
    condition: str
    is_active: bool
    message: str
    timestamp: str


class AlertsResponse(BaseModel):
    active_alerts_count: int
    active_alerts: List[AlertConditionItem]
    resolved_alerts: List[AlertConditionItem]
    configured_thresholds: Dict[str, Any]


class DriftMonitoringResponse(BaseModel):
    monitoring_notice: str
    overall_drift_status: str
    total_features_evaluated: int
    stable_features_count: int
    drifted_features_count: int
    features: List[Dict[str, Any]]


class ObservabilitySummaryResponse(BaseModel):
    system_health: SystemHealthItem
    runtime_metrics: RuntimeMetricsResponse
    forecast_operations: ForecastOperationsMetrics
    model_integrity_status: str
    dataset_integrity_status: str
    strategy_registry_status: str
    active_alerts_count: int
    timestamp: str
